third_party/move/move-vm/runtime/src/storage/struct_name_index_map.rs

```diff
@@ -42,59 +42,58 @@ impl StructNameIndexMap {
     /// if A == B, they have the same indices.
     pub(crate) fn struct_name_to_idx(
         &self,
-        struct_name: StructIdentifier,
+        struct_name: &StructIdentifier,
     ) -> PartialVMResult<StructNameIndex> {
-        // Note that we take a write lock here once, instead of (*): taking a read lock, checking
-        // if the index is cached, re-acquiring the (write) lock, and checking again, as it makes
-        // things faster.
-        // Note that even if we do (*), we MUST check if another thread has cached the index before
-        // we reached this point for correctness. If we do not do this, we can end up evicting the
-        // previous index, and end up with multiple indices corresponding to the same struct. As
-        // indices are stored inside types, type comparison breaks!
-        let mut index_map = self.0.write();
-
-        // Index is cached, return early.
-        if let Some(idx) = index_map.forward_map.get(&struct_name) {
-            return Ok(StructNameIndex(*idx));
+        {
+            let index_map = self.0.read();
+            if let Some(idx) = index_map.forward_map.get(struct_name) {
+                return Ok(StructNameIndex(*idx));
+            }
         }
 
-        // Otherwise, the cache is locked and the struct name is not present. We simply add it
-        // to the cache and return the corresponding index.
-        let idx = index_map.backward_map.len();
-        let prev_idx = index_map.forward_map.insert(struct_name.clone(), idx);
-        index_map.backward_map.push(Arc::new(struct_name));
+        // Possibly need to insert, so make the copies outside of the lock
+        let forward_key = struct_name.clone();
+        let backward_value = Arc::new(struct_name.clone());
 
-        // Unlock the cache.
-        drop(index_map);
+        let idx = {
+            let mut index_map = self.0.write();
+
+            if let Some(idx) = index_map.forward_map.get(struct_name) {
+                return Ok(StructNameIndex(*idx));
+            }
+
+            let idx = index_map.backward_map.len();
+            index_map.backward_map.push(backward_value);
+            index_map.forward_map.insert(forward_key, idx);
+            idx
+        };
 
-        if prev_idx.is_some() {
```

Comment:
Removed this as it can never happen - the `Some(idx)` case is dealt with under the same write lock!

Reply:
This is more of a paranoid check, but yes, tis is unreachable so removing should be ok